docs/pallas/design.md

```diff
@@ -0,0 +1,317 @@
+
+#### Reading from/writing into Refs
+
+Reading from a `Ref` corresponds to loading an array into the lowest level of the memory hierarchy (L1-cache on GPU and vector registers on TPU). Writing into a `Ref` is analogous.
+
+```python
+def f(x_ref, o_ref):
+  # Using vanilla Python indexing
+  x = x_ref[0, 2:5, :]
+  # Or via Numpy advanced int indexing
+  o_ref[jnp.arange(3), :] = x
+
+Note that in order to use NumPy advanced int indexing, you need to broadcast the indices against each other into the desired multidimensional shape:
+def f(x_ref):
+  # Assume x_ref is (8, 4) and we want to read out a (2, 3) slice
+  x = x_ref[jnp.arange(2)[..., None], jnp.arange(3)[None, ...]]
+```
+
+Writing to `Ref`s can be done via analogous `__setitem__` style indexing.
+
+Other forms of indexing (for example, dynamic slicing) can be done via `pallas.load` and `pallas.store`, new JAX primitives designed to make loading from/storing into memory easier. We’ll discuss these new primitives later.
+
+### Extending JAX with new Pallas primitives
+
+Because JAX was designed with HLO in mind, the set of JAX primitives closely mirrors the set of HLO operations. Targeting a new compiler (e.g. Triton or Mosaic) means we might need to supplement JAX’s primitives with new ones specific to the new compiler. At the same time, we may not be able to lower all JAX primitives, so we need to restrict it to a subset.
+
+Because Pallas was initially designed with Triton in mind, we offer a set of new primitives with the Triton programming model in mind. As we’ll show later, we can lower these primitives to Mosaic as well.
+
+#### `pallas.load` and `pallas.store`
+`pallas.load` and `pallas.store` are primitives that allow loading from memory and storing into memory. Unlike `__getitem__` and `__setitem__` they are more flexible at the cost of being more verbose. Specifically, you can use the `pallas.dynamic_slice` (`pallas.ds` for short) construct (which should maybe be upstreamed into JAX to be used with Ref `__getitem_` and `__setitem__`).
+
+```python
+def f(x_ref, o_ref):
+  # Reading from memory via pallas.load
+  x = pl.load(x_ref, (0, slice(2, 5), slice(None)))
+  # Using integer indexing automatically broadcasts
+  x = pl.load(x_ref, (0, 2 + jnp.arange(3), slice(None)))
+  # You can also use `pl.dynamic_slice` (`pl.ds` for short) objects as well
+  pl.store(o_ref, (0, pl.ds(start=2, size=3), slice(None)), x)
+```
+
+
+`pallas.load` and `pallas.store` also support masking via the mask argument.
+
+```python
+def f(x_ref, o_ref):
+  # Reading from memory via pallas.load
+  idx = jnp.arange(8)
+  mask = idx < 5
+  x = pl.load(x_ref, (idx,), mask=mask, other=float('-inf'))
```

Comment:
Have you ever considered adopting JAX-style `at[].get()` syntax for this? e.g. you could imagine writing this `pl.load` statement as
```python
x = x_ref.at[idx].get(mask=mask, other=float('-inf'))
```

Reply:
We could do something like that, though I wanted to draw a contrast between slicing an array and loading from a `Ref`, since `Ref`s are not arrays. Maybe something like `x_ref.at[idx].load(...)` and `x_ref.at[idx].store(...)`.

Reply:
Yea I think that's a good idea @froystig @mattjj

Reply:
Can you summarize the upside, e.g. with an example?

Reply:
IIUC it's roughly that we'd replace `lax.dynamic_slice(x, [idx], [size])` with `x[ds(idx, size)]`?